File: app/src/lib/identity/settlement-lineage.ts
/**
 * Settlement lineage — every settled result must be reconstructable end to end.
 *
 * WHY THIS EXISTS, CONCRETELY
 * The Sprint 044 audit traced three historical event-identity collisions through to settlement and
 * found 49 settled legs carrying a WRONG outcome. All three collisions were doubleheaders; in each the
 * surviving `gamePk` was game 2, so game 1's predictions were graded against game 2's box score.
 * Because both halves of a doubleheader share rosters, those legs graded to Win/Loss rather than
 * erroring — the failure was plausible, not missing, which is why it survived for months.
 *
 * The lesson is not "check for doubleheaders". It is that a settled result which cannot be traced back
 * through prediction → event → market → source is unfalsifiable. Nobody can tell a correct result from
 * a confidently wrong one without the chain, and an unfalsifiable win rate is not evidence.
 *
 * So this validator requires the whole chain to be present and internally consistent, and fails on the
 * specific shapes that produced the 49 bad legs.
 *
 * Sport-independent: no imports from any sport's modules.
 */
import { buildAliasIndex } from "./event-identity";

/** One settled result together with the chain that produced it. */
export interface SettlementLineage {
  /** The prediction being settled. Unique per settled row. */
  readonly predictionId: string;
  /** OUR canonical event id — never a provider id. */
  readonly eventId: string;
  /** The market the prediction was made against. */
  readonly marketId: string;
  readonly outcome: string;
  /** Where the result came from — an official box score, never a model's own output or a web snippet. */
  readonly settlementSource: string;
  /** When settlement ran. Must not precede the event. */
  readonly settledAt: string;
  /** When the event started, for the ordering check. Null makes ordering unprovable, not fine. */
  readonly eventStart?: string | null;
  /** The provider identifier the settlement actually joined on, when it used one. */
  readonly joinedProviderId?: string | null;
}

export type LineageViolationCode =
  /** A required link in the chain is absent. */
  | "MISSING_LINEAGE"
  /** Two settled rows claim the same predictionId. */
  | "DUPLICATE_PREDICTION"
  /** One provider id was joined to more than one event — the defect that produced the 49 bad legs. */
  | "DUPLICATE_MAPPING"
  /** The settlement joined on a provider id no event claims. */
  | "UNRESOLVED_PROVIDER"
  /** Settled before the event started, or another impossible ordering. */
  | "IMPOSSIBLE_RELATIONSHIP"
  /** The result came from something that is not an official source. */
  | "UNTRUSTED_SOURCE";

export interface LineageViolation {
  readonly code: LineageViolationCode;
  readonly message: string;
  readonly subjects: readonly string[];
}

/**
 * Sources that may settle a result.
 *
 * An allowlist rather than a denylist: a source nobody thought to forbid is exactly the one that ends
 * up settling a leg from a search-result snippet. Adding an entry here should require justifying it.
 */
export const OFFICIAL_SETTLEMENT_SOURCES: readonly string[] = [
  "mlb-statsapi-boxscore",
  "mlb-statsapi-linescore",
  "nba-stats-boxscore",
  "api-football-fixtures",
  "espn-official-scores",
  "operator-official-input",
];

const parse = (v: string | null | undefined): number | null => {
  if (!v) return null;
  const t = Date.parse(v);
  return Number.isFinite(t) ? t : null;
};

/**
 * Validate a set of settled results.
 *
 * Returns every violation rather than stopping at the first — a settlement run wants the full picture,
 * and reporting one problem at a time turns a single audit into six.
 */
export function validateSettlementLineage(
  rows: readonly SettlementLineage[],
): LineageViolation[] {
  const violations: LineageViolation[] = [];

  // 1 — every link present. A chain missing a link cannot be checked further, so these rows are
  //     reported and then excluded from the relational checks below rather than crashing them.
  const structural = new Set<string>();
  for (const r of rows) {
    const missing = (
      [
        ["predictionId", r.predictionId],
        ["eventId", r.eventId],
        ["marketId", r.marketId],
        ["outcome", r.outcome],
        ["settlementSource", r.settlementSource],
        ["settledAt", r.settledAt],
      ] as const
    )
      .filter(([, v]) => !v)
      .map(([k]) => k);
    if (missing.length > 0) {
      structural.add(r.predictionId || "(no predictionId)");
      violations.push({
        code: "MISSING_LINEAGE",
        message: `settled row "${r.predictionId || "(no predictionId)"}" is missing ${missing.join(", ")} — the result cannot be reconstructed`,
        subjects: [r.predictionId || "(no predictionId)"],
      });
    }
  }
  const wellFormed = rows.filter((r) => !structural.has(r.predictionId || "(no predictionId)"));

  // 2 — one settled row per prediction.
  const seen = new Map<string, number>();
  for (const r of wellFormed) seen.set(r.predictionId, (seen.get(r.predictionId) ?? 0) + 1);
  for (const [id, n] of seen) {
    if (n > 1) {
      violations.push({
        code: "DUPLICATE_PREDICTION",
        message: `prediction "${id}" was settled ${n} times — one of them is graded against the wrong event or double-counted`,
        subjects: [id],
      });
    }
  }

  // 3 — a provider id must map to exactly one event. THIS is the 49-bad-legs check.
  const joins = wellFormed.flatMap((r) =>
    r.joinedProviderId ? [[r.joinedProviderId, r.eventId] as const] : [],
  );
  const index = buildAliasIndex<string>(joins);
  for (const providerId of index.ambiguousAliases) {
    const events = [...new Set(joins.filter(([a]) => a === providerId).map(([, e]) => e))];
    violations.push({
      code: "DUPLICATE_MAPPING",
      message: `provider id "${providerId}" was settled against ${events.length} different events (${events.join(", ")}) — at least one set of results is graded against the wrong event`,
      subjects: [providerId, ...events],
    });
  }

  // 4 — timing and source.
  for (const r of wellFormed) {
    const settledAt = parse(r.settledAt);
    if (settledAt == null) {
      violations.push({
        code: "IMPOSSIBLE_RELATIONSHIP",
        message: `settled row "${r.predictionId}" has an unparseable settledAt "${r.settledAt}"`,
        subjects: [r.predictionId],
      });
    } else {
      const start = parse(r.eventStart);
      if (start != null && settledAt < start) {
        violations.push({
          code: "IMPOSSIBLE_RELATIONSHIP",
          message: `settled row "${r.predictionId}" was settled at ${r.settledAt}, BEFORE its event started at ${r.eventStart} — the outcome did not exist yet`,
          subjects: [r.predictionId],
        });
      }
    }

    if (!OFFICIAL_SETTLEMENT_SOURCES.includes(r.settlementSource)) {
      violations.push({
        code: "UNTRUSTED_SOURCE",
        message: `settled row "${r.predictionId}" cites source "${r.settlementSource}", which is not an official settlement source`,
        subjects: [r.predictionId, r.settlementSource],
      });
    }
  }

  return violations;
}

/**
 * Cross-check settled rows against the events that are actually known.
 *
 * Kept separate from `validateSettlementLineage` because it needs the event universe, which a caller
 * mid-settlement may not have. A settled row pointing at an unknown event is how an orphan hides.
 */
export function validateAgainstKnownEvents(
  rows: readonly SettlementLineage[],
  knownEventIds: readonly string[],
): LineageViolation[] {
  const known = new Set(knownEventIds);
  return rows
    .filter((r) => r.eventId && !known.has(r.eventId))
    .map((r) => ({
      code: "UNRESOLVED_PROVIDER" as const,
      message: `settled row "${r.predictionId}" points at event "${r.eventId}", which no known event claims`,
      subjects: [r.predictionId, r.eventId],
    }));
}

/**
 * Refuse to publish settled results that cannot be reconstructed.
 *
 * Throws. A settlement run whose lineage is broken should stop, not annotate — the whole reason the 49
 * bad legs persisted is that nothing downstream treated a suspicious result as a failure.
 */
export function assertSettlementPublishable(violations: readonly LineageViolation[]): void {
  if (violations.length === 0) return;
  const lines = violations.map((v) => `  [${v.code}] ${v.message}`).join("\n");
  throw new Error(
    `Settlement lineage validation failed — refusing to publish:\n${lines}\n\n` +
      `Every settled result must trace prediction → event → market → official source. ` +
      `See lib/identity/settlement-lineage.ts.`,
  );
}

File: pipeline/mlb/settlement_lineage.py
"""Settlement lineage enforcement for MLB — the Python side of the contract.

SPRINT 045. Sprint 044 proved three historical event-identity collisions corrupted 49 settled legs:
all three were doubleheaders where the late game's ``gamePk`` survived a last-write-wins join, so game
1's predictions were graded against game 2's box score. Because both halves share rosters, those legs
graded to Win/Loss instead of erroring — plausible, not missing, which is why nobody investigated.

Sprint 044 also built a validator for exactly that shape. It was a *test*, not enforcement: it could
prove a defect was catchable, and it sat in `app/src/lib/identity/settlement-lineage.ts` where the
Python settlement pipeline could never call it. This module closes that gap.

WHY A SECOND IMPLEMENTATION RATHER THAN A SHARED ONE
Settlement is Python; the surfaces are TypeScript. Bridging them at runtime would mean a subprocess
call inside the ledger write path — more failure modes than the check removes. So the contract is
mirrored here, and `settlement_lineage_test.py` asserts the two implementations agree on shared
fixtures, including the real 2026-07-22 collision. Two independent derivations that agree is stronger
evidence than one shared implementation nobody re-checks — the same argument that made the Sprint 043
historical audit credible.

Scope: pure functions over already-built rows. No I/O, no network.
"""
from __future__ import annotations

import re
import unicodedata
from datetime import datetime


class SettlementLineageError(RuntimeError):
    """Raised when settled rows would be written without a reconstructable lineage."""


# Mirrors OFFICIAL_SETTLEMENT_SOURCES in app/src/lib/identity/settlement-lineage.ts.
# An ALLOWLIST, not a denylist: a source nobody thought to forbid is exactly the one that ends up
# settling a leg from a search-result snippet.
OFFICIAL_SETTLEMENT_SOURCES = (
    "mlb-statsapi-boxscore",
    "mlb-statsapi-linescore",
    "nba-stats-boxscore",
    "api-football-fixtures",
    "espn-official-scores",
    "operator-official-input",
)

REQUIRED_FIELDS = ("id", "eventId", "marketKey", "outcome", "settlementSource", "settledAt")


def _slug(value: str) -> str:
    """Mirror of the TS `slug` helper. Must stay byte-compatible — the agreement test proves it."""
    norm = unicodedata.normalize("NFKD", value)
    norm = "".join(c for c in norm if not unicodedata.combining(c))
    norm = norm.lower()
    norm = re.sub(r"[^a-z0-9]+", "-", norm)
    return norm.strip("-")


def derive_event_id(
    *,
    sport: str,
    league: str | None,
    participant_names: list[str],
    scheduled_start: str | None,
) -> str:
    """Mirror of `deriveEventId` in app/src/lib/identity/event-identity.ts.

    Includes the start time to the MINUTE, which is what separates the two halves of a doubleheader.
    Date alone is what failed in Sprint 041. Participants are sorted so argument order cannot change
    the id — two adapters describing the same event must agree or the whole point is lost.
    """
    parts = sorted(_slug(n) for n in participant_names)
    if scheduled_start:
        when = re.sub(r":\d{2}(\.\d+)?Z?$", "", scheduled_start)
        when = when.replace(":", "").replace("-", "").replace("T", "t")
    else:
        when = "unscheduled"
    sport_slug = _slug(sport)
    league_slug = _slug(league) if league else None
    segments = [sport_slug, league_slug] if league_slug and league_slug != sport_slug else [sport_slug]
    return ":".join([*segments, "-v-".join(parts), when])


def _parse(value) -> datetime | None:
    if not value or not isinstance(value, str):
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None


def validate_settlement_lineage(rows: list[dict]) -> list[str]:
    """Return every lineage violation in a batch of settled rows. Empty means writable.

    Violations are returned rather than raised so a settlement run reports all of them at once —
    reporting one problem per run turns a single audit into six.
    """
    violations: list[str] = []

    # 1 — every link present. Rows failing this are excluded from the relational checks below rather
    #     than crashing them, so one malformed row cannot mask the rest.
    malformed: set[int] = set()
    for i, r in enumerate(rows):
        missing = [f for f in REQUIRED_FIELDS if not r.get(f)]
        if missing:
            malformed.add(i)
            violations.append(
                f"MISSING_LINEAGE: settled row {r.get('id') or '(no id)'} is missing "
                f"{', '.join(missing)} — the result cannot be reconstructed"
            )
    well_formed = [r for i, r in enumerate(rows) if i not in malformed]

    # 2 — one settled row per prediction.
    counts: dict[str, int] = {}
    for r in well_formed:
        counts[r["id"]] = counts.get(r["id"], 0) + 1
    for pid, n in sorted(counts.items()):
        if n > 1:
            violations.append(
                f"DUPLICATE_PREDICTION: prediction {pid} was settled {n} times — one of them is "
                f"graded against the wrong event or double-counted"
            )

    # 3 — THE 49-BAD-LEGS CHECK. A provider event id and a canonical event must be one-to-one, in
    #     BOTH directions. One provider id across two events means some rows are graded against the
    #     wrong game; one event across two provider ids means the join is ambiguous.
    by_provider: dict[str, set[str]] = {}
    by_event: dict[str, set[str]] = {}
    for r in well_formed:
        alias = r.get("providerEventId") or r.get("gameId")
        if not alias:
            continue
        by_provider.setdefault(alias, set()).add(r["eventId"])
        by_event.setdefault(r["eventId"], set()).add(alias)
    for alias, events in sorted(by_provider.items()):
        if len(events) > 1:
            violations.append(
                f"DUPLICATE_MAPPING: provider id {alias[:16]} was settled against {len(events)} "
                f"different events ({', '.join(sorted(events))}) — at least one set of results is "
                f"graded against the wrong event"
            )
    for event, aliases in sorted(by_event.items()):
        if len(aliases) > 1:
            violations.append(
                f"AMBIGUOUS_IDENTITY: event {event} was settled from {len(aliases)} different "
                f"provider ids ({', '.join(sorted(a[:16] for a in aliases))})"
            )

    # 4 — a canonical event must resolve to exactly one gamePk. This is the collision expressed in the
    #     settlement's own terms, and it is what would have caught 2026-07-22 at write time.
    pks_by_event: dict[str, set] = {}
    for r in well_formed:
        if r.get("gamePk") is not None:
            pks_by_event.setdefault(r["eventId"], set()).add(r["gamePk"])
    for event, pks in sorted(pks_by_event.items()):
        if len(pks) > 1:
            violations.append(
                f"WRONG_EVENT_MAPPING: event {event} was graded against {len(pks)} gamePks "
                f"({', '.join(str(p) for p in sorted(pks, key=str))})"
            )

    # 4b — the INVERSE, and the one that actually fires on the real 2026-07-22 board: two distinct
    #      canonical events graded against the SAME gamePk. The eventIds correctly differ there
    #      (derivation separates the halves by start time), so checks 3 and 4 pass — but both halves
    #      still carried gamePk 823519, which is precisely how game 1's legs were graded against game
    #      2's box score. Found by running this gate against the real board rather than a fixture.
    events_by_pk: dict = {}
    for r in well_formed:
        if r.get("gamePk") is not None:
            events_by_pk.setdefault(r["gamePk"], set()).add(r["eventId"])
    for pk, events in sorted(events_by_pk.items(), key=lambda kv: str(kv[0])):
        if len(events) > 1:
            violations.append(
                f"WRONG_EVENT_MAPPING: gamePk {pk} is claimed by {len(events)} distinct events "
                f"({', '.join(sorted(events))}) — one event's predictions would be graded against "
                f"the other's box score"
            )

    # 5 — timing and source.
    for r in well_formed:
        settled_at = _parse(r.get("settledAt"))
        if settled_at is None:
            violations.append(
                f"IMPOSSIBLE_RELATIONSHIP: settled row {r['id']} has an unparseable settledAt "
                f"{r.get('settledAt')!r}"
            )
        else:
            start = _parse(r.get("eventStartTime"))
            if start is not None and settled_at < start:
                violations.append(
                    f"IMPOSSIBLE_RELATIONSHIP: settled row {r['id']} was settled at "
                    f"{r['settledAt']}, BEFORE its event started at {r['eventStartTime']} — "
                    f"the outcome did not exist yet"
                )
        if r.get("settlementSource") not in OFFICIAL_SETTLEMENT_SOURCES:
            violations.append(
                f"UNTRUSTED_SOURCE: settled row {r['id']} cites source "
                f"{r.get('settlementSource')!r}, which is not an official settlement source"
            )

    return violations


def assert_settlement_lineage(rows: list[dict], *, date: str) -> None:
    """Refuse to write settled rows whose lineage is broken.

    Raises rather than warning. The whole reason the 49 corrupted legs persisted is that nothing
    downstream treated a suspicious result as a failure — a warning here would be read exactly as
    every other warning in this pipeline's history was.
    """
    violations = validate_settlement_lineage(rows)
    if not violations:
        return
    detail = "\n".join(f"  {v}" for v in violations)
    raise SettlementLineageError(
        f"Settlement {date} failed lineage validation — refusing to write {len(rows)} row(s):\n"
        f"{detail}\n\n"
        f"  Every settled result must trace prediction -> event -> market -> official source.\n"
        f"  See pipeline/mlb/settlement_lineage.py and app/src/lib/identity/settlement-lineage.ts."
    )
